src/factorizations/blocklanczos.jl

```diff
@@ -0,0 +1,322 @@
+hermitian directly when the linear map is encoded as a general callable object or function, with `block_inner(X, f.(X))`,
+it is tested whether `norm(M-M')` is sufficiently small to be neglected.
+
+The argument `f` can be a matrix, or a function accepting a single argument `x`, so that
+`f(x)` implements the action of the linear map on the block `x`.
+
+The optional argument `orth` specifies which [`Orthogonalizer`](@ref) to be used. The
+default value in [`KrylovDefaults`](@ref) is to use [`ModifiedGramSchmidt2`](@ref), which
+uses reorthogonalization steps in every iteration.
+Now our orthogonalizer is only ModifiedGramSchmidt2. So we don't need to provide "keepvecs" because we have to reverse all krylove vectors.
+Dimension of Krylov subspace in BlockLanczosIterator is usually much bigger than lanczos and its Default value is 100.
+`qr_tol` is the tolerance used in [`abstract_qr!`](@ref) to resolve the rank of a block of vectors.
+
+When iterating over an instance of `BlockLanczosIterator`, the values being generated are
+instances of [`BlockLanczosFactorization`](@ref). 
+
+The internal state of `BlockLanczosIterator` is the same as the return value, i.e. the
+corresponding `BlockLanczosFactorization`.
+
+Here, [`initialize(::KrylovIterator)`](@ref) produces the first Krylov factorization,
+and [`expand!(iter::KrylovIterator, fact::KrylovFactorization)`](@ref) expands the
+factorization in place.
+"""
+struct BlockLanczosIterator{F,T,S,O<:Orthogonalizer} <: KrylovIterator{F,T}
+    operator::F
+    x₀::BlockVec{T,S}
+    maxdim::Int
+    orth::O
+    qr_tol::Real
+    function BlockLanczosIterator{F,T,S,O}(operator::F,
+                                           x₀::BlockVec{T,S},
+                                           maxdim::Int,
+                                           orth::O,
+                                           qr_tol::Real) where {F,T,S,O<:Orthogonalizer}
+        return new{F,T,S,O}(operator, x₀, maxdim, orth, qr_tol)
+    end
+end
+function BlockLanczosIterator(operator::F,
+                              x₀::BlockVec{T,S},
+                              maxdim::Int,
+                              qr_tol::Real,
+                              orth::O=ModifiedGramSchmidt2()) where {F,T,S,
+                                                                     O<:Orthogonalizer}
+    norm(x₀) < qr_tol && @error "initial vector should not have norm zero"
+    orth != ModifiedGramSchmidt2() &&
+        @error "BlockLanczosIterator only supports ModifiedGramSchmidt2 orthogonalizer"
+    return BlockLanczosIterator{F,T,S,O}(operator, x₀, maxdim, orth, qr_tol)
+end
+
+function initialize(iter::BlockLanczosIterator{F,T,S};
+                    verbosity::Int=KrylovDefaults.verbosity[]) where {F,T,S}
+    X₀ = iter.x₀
+    maxdim = iter.maxdim
+    bs = length(X₀) # block size now
+    A = iter.operator
+    TDB = zeros(S, maxdim, maxdim)
+
+    # Orthogonalization of the initial block
+    X₁ = deepcopy(X₀)
```

Comment:
Rather than relying on `deepcopy`, I would prefer to have a method
```julia
function Base.copy(b::BlockVec)
    return BlockVec(scale.(b.vec, 1))
end
```